File: Sean的leetcode/FB/Flatten Nested List Iterator.py
###
# Stack
# Time Complexity: next: O(1) hasNext:O(n)
# Space Complexity: O(n)
###
# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """
#class NestedInteger(object):
#    def isInteger(self):
#        """
#        @return True if this NestedInteger holds a single integer, rather than a nested list.
#        :rtype bool
#        """
#
#    def getInteger(self):
#        """
#        @return the single integer that this NestedInteger holds, if it holds a single integer
#        Return None if this NestedInteger holds a nested list
#        :rtype int
#        """
#
#    def getList(self):
#        """
#        @return the nested list that this NestedInteger holds, if it holds a nested list
#        Return None if this NestedInteger holds a single integer
#        :rtype List[NestedInteger]
#        """

class NestedIterator(object):

    def __init__(self, nestedList):
        """
        Initialize your data structure here.
        :type nestedList: List[NestedInteger]
        """
        self.stack = [[nestedList, 0]]

    def next(self):
        """
        :rtype: int
        """
        if self.hasNext():
            lst, i = self.stack[-1]
            self.stack[-1][1] += 1
            return lst[i].getInteger()
        else:
            return -1
        

    def hasNext(self):
        """
        :rtype: bool
        """
        s = self.stack
        while s:
            lst, i = s[-1]
            if i == len(lst):
                s.pop()
            else:
                e = lst[i]
                if e.isInteger():
                    return True
                s[-1][1] += 1
                s.append([e.getList(), 0])
        return False

# Your NestedIterator object will be instantiated and called as such:
# i, v = NestedIterator(nestedList), []
# while i.hasNext(): v.append(i.next())

File: MyLeetCode/Leetcode/Flatten Nested List Iterator.cpp
// Concept: use Stack for LIFO because we need to divide the inner list
// Iterator shouldn't copy the entire data but just iterate over the original data structure.
// keep the current progress in a stack. 
// hasNext tries to find an integer
// next returns it and moves on

/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * class NestedInteger {v
 *   public:
 *     // Return true if this NestedInteger holds a single integer, rather than a nested list.
 *     bool isInteger() const;
 *
 *     // Return the single integer that this NestedInteger holds, if it holds a single integer
 *     // The result is undefined if this NestedInteger holds a nested list
 *     int getInteger() const;
 *
 *     // Return the nested list that this NestedInteger holds, if it holds a nested list
 *     // The result is undefined if this NestedInteger holds a single integer
 *     const vector<NestedInteger> &getList() const;
 * };
 */
class NestedIterator {
public:
    NestedIterator(vector<NestedInteger> &nestedList) {
        begin.push(nestedList.begin());
        end.push(nestedList.end()); //point to the one after the last element
    }

    int next() {
        if (hasNext()){
            auto tmp = start.top();
            start.top()++;
            return tmp->getInteger();
        }
        return -1;
    }

    bool hasNext() {
        while(!begin.empty()){
            if(begin.top() == end.top()){
                begin.pop();
                end.pop();
            }
            else{
                auto x = begin.top();
                if(x->isInteger()){
                    return true;
                }
                begin.top()++;
                begin.push(x->getList().begin());
                end.push(x->getList().end());
            }
            
            
        }
        return false;
    }
private:
    stack<vector<NestedInteger>::iterator> begin, end;
};

/**
 * Your NestedIterator object will be instantiated and called as such:
 * NestedIterator i(nestedList);
 * while (i.hasNext()) cout << i.next();
 */